share/arno-iptables-firewall/plugins/20parasitic-net.plugin

```diff
@@ -0,0 +1,292 @@
+#
+PLUGIN_NAME="Parasitic Network plugin"
+PLUGIN_VERSION="1.00 BETA"
+PLUGIN_CONF_FILE="parasitic-net.conf"
+#
+# Last changed          : July 25, 2017
+# Requirements          : AIF 2.0.1+
+# Comments              : This plugin allows "clients" on the same subnet to use this
+#                       : device as a gateway upstream. This network of "clients" is
+#                       : SNAT'ed to this device's external interface(s).
+#                       : 
+#                       : This Parasitic Network is useful for situations when the
+#                       : upstream firewall is not under your control and you desire
+#                       : added security for specific devices in your subnet.
+#                       : Set the gateway address of Parasitic Network clients to an
+#                       : external IPv4 address of this device.
+#
+# Author                : (C) Copyright 2017 by Arno van Amersfoort & Lonnie Abelbeck
+# Email                 : arnova AT rocky DOT eld DOT leidenuniv DOT nl
+#                         (note: you must remove all spaces and substitute the @ and the .
+#                         at the proper locations!)
+# ------------------------------------------------------------------------------
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# version 2 as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+# ------------------------------------------------------------------------------
+
+parasitic_net_first_ipv4()
+{
+  ip -o addr show dev "$1" 2>/dev/null \
+    | awk '$3 == "inet" { split($4, field, "/"); print field[1]; nextfile; }'
+}
+
+parasitic_net_all_ipv4()
+{
+  ip -o addr show dev "$1" 2>/dev/null \
+    | awk '$3 == "inet" { split($4, field, "/"); print field[1]; }'
+}
+
+# Plugin start function
+plugin_start()
+{
+  local x host net eif gateway_if gateway_ip snat_if snat_ip snat_ifs_ips IFS
+
+  ip4tables -t nat -N PARASITIC_NET_SNAT 2>/dev/null
+  ip4tables -t nat -F PARASITIC_NET_SNAT
+
+  ip4tables -N PARASITIC_NET_ACL 2>/dev/null
+  ip4tables -F PARASITIC_NET_ACL
+
+  ip4tables -N PARASITIC_NET_FORWARD 2>/dev/null
+  ip4tables -F PARASITIC_NET_FORWARD
+
+  snat_ifs_ips=""
+  IFS=' ,'
+  for eif in $(wildcard_ifs $EXT_IF); do
+    snat_if="$eif"
+    snat_ip="$(parasitic_net_first_ipv4 $eif)"
+    if [ -n "$snat_if" -a -n "$snat_ip" ]; then
+      snat_ifs_ips="$snat_ifs_ips${snat_ifs_ips:+ }$snat_if~$snat_ip"
+    fi
+  done
+
```

Comment:
The `snat_ifs_ips` list of interface~IPv4 pairs is new in the latest commit.

For multiple `EXT_IF` interfaces, the `snat_ifs_ips` list contains all possible interfaces that many need to be SNAT'ed given any arbitrary route outbound.

The `snat_ifs_ips` list is used both when `PARASITIC_NET_IF` is set or unset.

If we wanted we could add a configuration variable to specify external interfaces that we don't want to SNAT to, though the extra SNAT rules can be needed for failover situations when the route is dynamically changed.